sycl/doc/extensions/experimental/sycl_ext_adaptivecpp_enqueue_custom_operation.asciidoc

```diff
@@ -0,0 +1,89 @@
+# sycl_ext_adaptivecpp_enqueue_custom_operation
+
+## Notice
+
+This document describes an **experimental** API that applications can use to try
+out a new feature. Future versions of this API may change in ways that are
+incompatible with this experimental version.
+
+## Introduction
+
+This extension is derived from an experimental AdaptiveCpp extension, which is
+documented
+[here](https://github.com/sycl_ext_oneapi/sycl_ext_oneapi/blob/develop/doc/enqueue-custom-operation.md)
+
+## Example
+
+```c++
+sycl::queue q;
+
+q.submit([&](sycl::handler &cgh) {
+    sycl::accessor acc{buf, cgh};
+
+    cgh.sycl_ext_oneapi_enqueue_custom_operation([=](sycl::interop_handle &h) {
+      // Can extract device pointers from accessors
+      void *native_mem = h.get_native_mem<sycl::backend::hip>(acc);
+      // Can extract stream (note: get_native_queue() may not be 
+      // supported on CPU backends)
+      hipStream_t stream = h.get_native_queue<sycl::backend::hip>();
+      // Can extract HIP device (note: get_native_device() may not be
+      // supported on CPU backends)
+      int dev = h.get_native_device<sycl::backend::hip>();
+      // Can enqueue arbitrary backend operations. This could also be a kernel launch
+      // or a call to a library that enqueues operations on the stream etc
+      hipMemcpyAsync(target_ptr, native_mem, test_size * sizeof(int),
+                      hipMemcpyDeviceToHost, stream);
+    });
+  });
+q.wait()
+```
+
+Provided that the correct dependencies are handled in the SYCL runtime, there
+is no need to do any further synchronization within the backend API in the user
+submitted lambda.
+
+## Synchronization
+
+The goal of `sycl_ext_oneapi_enqueue_custom_operation` is to integrate interop
+work within the asynchronous SYCL DAG. As such, the user defined lambda must
+only enqueue asynchronous work within the user lambda. Asynchronous work must
+only be submitted  to the native queue obtained from
+`interop_handle::get_native_queue`.
+
+A lambda submitted to sycl_ext_oneapi_enqueue_custom_operation will be executed
+in the host thread, without waiting for any dependent events to complete. This
```

Comment:
I think this description is maybe too implementation-specific. An implementation could potentially wait on dependent events, though this API allows commands to be enqueued without waiting for dependent commands to  complete first. Also it doesn't need to be executed in the host thread, it could still be deferred to another thread.

I think the important thing to specify is that the function must wait for any dependent commands to be enqueued but they don't need to have completed execution.

We should also specify that any necessary synchronisation within the SYCL runtime to be performed, such that a native queue is provided which can be enqueued to. For example an out-of-order queue may need to perform a barrier.

Reply:
Thanks have reworded. Let me know what you think

Reply:
Thanks, the updated draft looks really good. My only comment would be say something about how the SYCL RT must provide a native stream/queue which allows enqueueing native commands and satisfying the requirement that dependent commands have been enqueued. It's subtle but there are some cases such as using an out-of-order queue or enqueueing this API after a host task where an implementation will need to take care that dependencies are satisfied and the correct execution order is maintained. Perhaps this could be partially in an implementation notes section.

Reply:
I've sort of covered that case [here](https://github.com/intel/llvm/pull/14136/files#diff-5e294cad2d805dc8b60b5b1a5279670994e8e782ff05a5509f1925bca5cdfde2R132). All I've said is that the synchronisation is guaranteed in the backend API as long as the user submits async work to the backend queue returned by `interop_handle::get_native_queue`. LMKWYT